### ISSUE TITLE
[Truffle] remove_const doesn't completely clear a module's name

### ISSUE THREAD

**Issue by nirvdrum** [jruby/jruby#3858]
`rspec-support` has a JRuby-specific fix that exposes a flaw in our `remove_const` implementation. It basically replaces a class with a subclass by using `remove_const` to remove the previous definition. Normally, a class cannot subclass itself, but `remove_const` should make the previous definition anonymous, at which point it can be subclassed.

A reduced example is:

```
module RSpec
  module Support
    class MethodSignature
    end

    class MethodSignature < remove_const(:MethodSignature)
      def x
        puts "Called x"
      end
    end
  end
end

RSpec::Support::MethodSignature.new.x
```

In MRI, this prints "Called X". In JRuby+Truffle, this results in a `TypeError` with a message of ``Support': superclass mismatch for class RSpec::Support::MethodSignature`.

This is currently a blocker for many cases where rspec is used.

**Comment by nirvdrum** [comment]
Just as a general note, the JRuby bug that `rspec-support` is attempting to workaround is #2816, which similarly affects JRuby+Truffle (probably due to shared code).

**Comment by chrisseaton** [comment]
I think this is fixed by 066f5c25b9fa861dc0f9bcb61622302ae624c0df, 7a224a52c582da630436f529efddde32260e93ba, but the latter might be too specific and not really fix the problem.

**Comment by eregon** [comment]
`but remove_const should make the previous definition anonymous, at which point it can be subclassed.`
It does not AFAIK, but it does remove the constant so that it's not found when looking up if the class already exist or not.

**Comment by eregon** [comment]
This workaround is very confusing though as it produces twice 2 classes with the same name in the chain:
`[A, A, Object, PP::ObjectMixin, Kernel, BasicObject]`.
There is already specs for arity and parameters of attr_ method AFAIK, we should fix those.

**Comment by eregon** [comment]
The first workaround of executing superclass should be enough.
@nirvdrum Could you confirm this work, even with 7a224a52c582da630436f529efddde32260e93ba reverted?

**Comment by nirvdrum** [comment]
While having two classes with the same name in the ancestry chain is confusing, this is also what MRI does. My phrasing around the term "anonymous" was ambiguous. Benoit's follow-up is accurate.

I tried reverting 7a224a52c582da630436f529efddde32260e93ba and the sample code outputs the same result as with the commit present. It appears safe to revert.

**Comment by eregon** [comment]
Closed by 7ed9a364050fd9624b67373663e00fce1ae83bbb
